### PR TITLE
[core][distributed] accelerate distributed weight loading

### PR DESCRIPTION
**Pull request by youkaichao**
try to make sure process only reads what it needs.

prior to this pr, the model reads all the weight, and then truncate the tensor. it requires every process to read all the model weight, which can be slow when the model is large.

### CODE REVIEW

**Comment by youkaichao** [comment]
# how weight loading is handled in distributed inference previously

all processes will read all the weights from disk:

https://github.com/vllm-project/vllm/blob/56b325e977435af744f8b3dca7af0ca209663558/vllm/model_executor/model_loader/weight_utils.py#L353-L362

Note the `f.get_tensor` call will return a tensor.

and then, in pipeline parallel case, we replace unused layer in the current stage as `nn.Identity()`, and suppress key error during weight loading:

image

in tensor parallel case, we truncate the loaded tensor and copy part of it to GPU.

in short, every process reads all the weights from disk, and then copy the necessary part to GPU. **all processes read the model weight for tp x pp  times in total.**

# how weight loading is handled in distributed inference in this PR

all processes will read all the weights **metadata** from disk:

image

Note the `DeferredTensor(f.get_slice(name))` call will return a **DeferredTensor**. only when we materialize it, will it read from disk.

 in pipeline parallel case, because key error happens early, we will not execute `weight_loader`. Therefore, `DeferredTensor` is not materialized if a layer does not appear in the current stage.

in tensor parallel case, we can truncate (narrow) the `DeferredTensor`, and then load it by calling `ensure_tensor()`. It will only load the part it needs from disk.

in short, **all process reads the weights from disk for just once**, and then copy their corresponding parts to GPU.

The benefit would be more significant, when the model becomes large.

**Comment by njhill** [comment]
@youkaichao this looks great! What kind of speedup does this give to e.g. mixtral or llama 70b?

How does it compare to https://github.com/vllm-project/vllm/pull/3729? In the TP case it might be beneficial for each worker to read sequentially from disjoint contiguous regions?

File: vllm/utils.py [hunk 1]
```diff
@@ -922,3 +922,35 @@ def parse_args(self, args=None, namespace=None):
                 processed_args.append(arg)
 
         return super().parse_args(processed_args, namespace)
+
+
+class DeferredTensor:
```
Comment (andoorve) [inline vllm/utils.py:927]: Why isn't it possible to subclass/use Meta tensor for example?

Comment (youkaichao) [reply vllm/utils.py:927]: Meta tensor is used to respond metadata related query now.

**Comment by youkaichao** [comment]
@njhill I think the benefit of this PR should be similar to #3729 , but with the current implementation, the user-facing changes will be minimal.

**Comment by robertgshaw2-neuralmagic** [comment]
This is awesome

File: vllm/utils.py [hunk 2]
```diff
@@ -922,3 +922,215 @@ def parse_args(self, args=None, namespace=None):
...
+        return +self.materialize()
+
+    def __abs__(self):
+        return abs(self.materialize())
+
+    def __invert__(self):
+        return ~self.materialize()
+
+
+def ensure_tensor(x: Union[DeferredTensor, torch.Tensor]) -> torch.Tensor:
```
Comment (andoorve) [inline vllm/utils.py:1132]: Looks like this doesn't get used

**Review by andoorve — APPROVED** [review]
LGTM, minimally intrusive change

As @njhill suggested would it be possible to get a sanity check on speedup for LLaMa 70B for example?

File: vllm/utils.py [hunk 3]
```diff
@@ -922,3 +922,215 @@ def parse_args(self, args=None, namespace=None):
...
+            return args[0].narrow(**kwargs)
+        new_args = []
+        for arg in args:
+            if isinstance(arg, DeferredTensor):
+                new_args.append(arg.materialize())
+            else:
+                new_args.append(arg)
+        return func(*new_args, **kwargs)
+
+    # implement common tensor operations, except for in-place operations
```
Comment (Yard1) [inline vllm/utils.py:1021]: why do we need those in the first place?

Comment (youkaichao) [reply vllm/utils.py:1021]: I write these code so that the deferred tensors can behave just like normal tensors. for example, some users call `torch.nn.functional.normalize` on tensors loaded from disk : 

https://github.com/vllm-project/vllm/blob/543aa4857362ff385af8a9f496392b8831c42833/vllm/model_executor/models/baichuan.py#L382

When we have these functions, users' code will be minimally affected.

File: vllm/utils.py (hunk 2 above)
Comment (youkaichao) [reply vllm/utils.py]: nice catch, deleted

File: vllm/utils.py (hunk 3 above)
Comment (Yard1) [reply vllm/utils.py:1023]: I think this will lead to issues later down the line (we now have an ill-defined subset of supported operations). I feel like explicitly requiring materialization is a better way.

Comment (youkaichao) [reply vllm/utils.py:1023]: you can refer to [da9d486](https://github.com/vllm-project/vllm/pull/6127/commits/da9d486cdc4a8d58133b1a7f0aec2332436f67fe) for example. that is what i did before. I find that code is more intrusive, and all third-party code using vllm might break.

**Comment by youkaichao** [comment]
strangely, I don't see significant speedup from myside. maybe it is because I use a machine with 2TB memory, and the OS is smart enough to cache the disk read?

**Comment by robertgshaw2-neuralmagic** [comment]
> strangely, I don't see significant speedup from myside. maybe it is because I use a machine with 2TB memory, and the OS is smart enough to cache the disk read?

I think the logic for weight loading from safetensors is running in the main process --- we pass the iterator to each worker [here](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/model_loader/loader.py#L271)

Perhaps we were already doing the right thing here

**Comment by njhill** [comment]
> strangely, I don't see significant speedup from myside. maybe it is because I use a machine with 2TB memory, and the OS is smart enough to cache the disk read?

@youkaichao you definitely need to clear the OS page cache between attempts when doing measurements

File: vllm/utils.py (hunk 3 above)
Comment (Yard1) [reply vllm/utils.py:1023]: I think this is too magical and being explicit would be better, but I am not going to block on this.

**Comment by youkaichao** [comment]
> I think the logic for weight loading from safetensors is running in the main process --- we pass the iterator to each worker [here](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/model_loader/loader.py?rgh-link-date=2024-07-08T20%3A22%3A24Z#L271)

this is not true. every process just get the config, and the rest is done separately .

**Comment by robertgshaw2-neuralmagic** [comment]
> > I think the logic for weight loading from safetensors is running in the main process --- we pass the iterator to each worker [here](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/model_loader/loader.py?rgh-link-date=2024-07-08T20%3A22%3A24Z#L271)
> 
> this is not true. every process just get the config, and the rest is done separately .

Yeah youre right on second glance.

**Comment by chestnut-Q** [comment]
> strangely, I don't see significant speedup from myside. maybe it is because I use a machine with 2TB memory, and the OS is smart enough to cache the disk read?

```Python
def convert_pyslice_to_tensor(
        x: Any, 
        slice_args: Optional[Tuple[int, int, int]] = None,
    ) -> torch.Tensor:
    """convert PySafeSlice object from safetensors to torch.Tensor
    """
    if slice_args is None:
        if not isinstance(x, torch.Tensor):
            x = x[:]
    else:
        dim, start, length = slice_args
        if not isinstance(x, torch.Tensor):
            slices = tuple([slice(start, start + length) 
                            if i == dim else slice(None) 
                            for i in range(len(x.get_shape()))])
            x = x[slices]
        else:
            x = x.narrow(dim, start, length)
    return x

for i in range(1, 6):
    with safe_open(f"{model_dir}/model-000{i:02d}-of-00015.safetensors", framework="pt") as f:
        for key in f.keys():
            tensor_a = f.get_slice(key)
            tensor_dict[key] = torch.zeros(tuple(tensor_a.get_shape()), dtype=torch.float16, device="cuda")

start = time.time()

for i in range(1, 6):
    with safe_open(f"{model_dir}/model-000{i:02d}-of-00015.safetensors", framework="pt") as f:
        for key in f.keys():
            
            tensor_a = f.get_tensor(key) # get_tensor
            # tensor_a = f.get_slice(key) # get_slice

            length = tensor_dict[key].shape[0] // 8
            tensor_dict[key][:length].copy_(convert_pyslice_to_tensor(tensor_a, (0, 0, length)))

end = time.time()
```
Hello @youkaichao, I tested the loading speeds of `get_tensor` and `get_slice` using the code snippet above, for the Llama-2-70B-fp16 model. The results indicate that despite `get_tensor` and `get_slice` theoretically involving an 8-fold difference in I/O volume, their actual speeds show no significant divergence. Further experiments revealed that `convert_pyslice_to_tensor` takes very little time compared to `copy_`, suggesting that neither `get_tensor` nor `get_slice` actually perform data copying in practice. During `copy_` (h2d), both methods transfer the same amount of data, thus explaining the minimal observed difference.

**Comment by youkaichao** [comment]
@chestnut-Q thanks for the report. what's your disk io speed and cpu memory?

**Comment by chestnut-Q** [comment]
> @chestnut-Q thanks for the report. what's your disk io speed and cpu memory?

1TB cpu memory and ~5GB/s disk io

**Comment by youkaichao** [comment]
measurement here should be careful, and you need to take care of OS cache (caching disk read).

**Comment by youkaichao** [comment]
we can suspend this PR until https://github.com/huggingface/safetensors/issues/496 is addressed. I find `get_tensor` and `get_slice` performs quite similar.

**Comment by youkaichao** [comment]
> During copy_ (h2d)

`copy_` here is complicated. because of `mmap`, when cuda driver tries to transfer this part of memory to device, OS will first read from disk, and hand it over to cuda driver.

That said, the `copy_` here involves both disk read and h2d transfer.

**Comment by sdake** [comment]
cc/ @sdake

File: vllm/utils.py (hunk 3 above)
Comment (sdake) [reply vllm/utils.py:1043]: @Yard1 would you consider elaborating. I am not sure I agree that explicitly presenting the tensor `bytearrray` content in a person's mental model ordered wrt consumption of said `bytearray` is wise, especially for sarge a large read-only buffer. I am of opinion that any large read-only memory is a massive target for performance optimization.

I have been working on a very small PR designed to improve availability. This is done by eliminating the model load startup delay (about 1 minute on my gear, I can shave it to about 3 seconds). This is pretty straightforward to deliver, presuming that the logit operation does not need to be calculated on startup.

The gist of my WIP PR is:
- remove logits calculation (which takes 12 seconds on my very beast GPU server with 2 A40 running llama3-70b-w8a8.
- Rely on safetensors exclusively.
- Modify, slightly, the management of pytorch `storage` such that IF a `storage` buffer address is accessed with any Tensor.to('cuda'), there will be some intelligence as it relates to h2d memcpy..
- Modify, slightly, to('cuda') calls to send Tensor objects to cuda devices, but not to execute a cuda host-to-device copy (as is default with Pytorch).

Would be interested in hearing your opinions on the approach and a review when ready. There are some constraints:
- safetensors first implementation
- cuda first implementation
- cuda must have `uvm` loaded
- dedicated address space consumption of model size
- dedicated (initially) memory consumption of model size (this is an obnoxious constraint, and is not permanent).
- no run-time quantization. (I prefer static compressed models, although I understand the value of some types of dynamic calculations of decompression, so we can kick this around).

One problem I struggle with is not technical - its about making this model an optional feature (via runtime environment variable), or adding as a flag.

Thanks that was helpful writing down my ideas in this area, I will copy this into an issue tracker. 
cc/ @sdake
cc/ @robertgshaw2-neuralmagic
